### PR TITLE
FileStore: fiemap implementation using SEEK_HOLE/SEEK_DATA

### PR DESCRIPTION
**Pull request by wonzhq**
Signed-off-by: Zhiqiang Wang <zhiqiang.wang@intel.com>

### CODE REVIEW

File: src/os/FileStore.cc [hunk 1]
```diff
@@ -2889,69 +2889,142 @@ int FileStore::fiemap(coll_t cid, const ghobject_t& oid,
...
+    off_t tmp_pos;
+    tmp_pos = lseek(**fd, 0, SEEK_SET);
+    if (tmp_pos != 0) {
+      r = -errno;
       goto done;
     }
+  }
+
+  if (!fallback_to_fiemap) {
+    // If lseek fails with errno setting to be ENXIO, this means the current
```
Comment (liewegas) [inline src/os/FileStore.cc]: ENXIO or EINVAL?

**Comment by liewegas** [comment]
It would probably simplify this a bit to make a couple of easy helpers like
<pre>
 int _do_fiemap(int fd, map&lt;uint64_t> *m)
 int _do_seek_hole_data(int fd, map&lt;uint64_t> *m)
</pre>
and have the caller try one then the other if it gets an error code.

File: src/os/FileStore.cc (hunk 1 above)
Comment (wonzhq) [inline src/os/FileStore.cc]: From the man page of lseek, the error of ENXIO means: whence is SEEK_DATA or SEEK_HOLE, and the current file offset is beyond the end of the file. So I think it's ENXIO here.

**Comment by wonzhq** [comment]
@liewegas , updated, pls take another look.

File: src/os/FileStore.cc [hunk 2]
```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
...
 
-      dout(10) << "FileStore::fiemap() fm_mapped_extents=" << fiemap->fm_mapped_extents
-	       << " fe_logical=" << extent->fe_logical << " fe_length=" << extent->fe_length << dendl;
+  // If compiled on an OS with SEEK_HOLE/SEEK_DATA support, but running
+  // on an OS that doesn't support SEEK_HOLE/SEEK_DATA, EINVAL is returned.
+  // Fall back to use fiemap.
+  hole_pos = lseek(fd, 0, SEEK_HOLE);
+  if (hole_pos < 0) {
+    if (errno == EINVAL) {
+      *fallback_to_fiemap = true;
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: log here

File: src/os/FileStore.cc [hunk 3]
```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
...
+  // If lseek fails with errno setting to be ENXIO, this means the current
+  // file offset is beyond the end of the file.
+  off_t start = offset;
+  while(start < (off_t)(offset + len)) {
+    data_pos = lseek(fd, start, SEEK_DATA);
+    if (data_pos < 0) {
+      if (errno == ENXIO)
+        break;
+      else {
+        r = -errno;
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: log here?

File: src/os/FileStore.cc [hunk 4]
```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
...
     }
-    free(fiemap);
+
+    if (hole_pos > (off_t)(offset + len)) {
+      (*m)[data_pos] = offset + len - data_pos;
+      break;
+    }
+    (*m)[data_pos] = hole_pos - data_pos;
+    start = hole_pos;
   }
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: log?

File: src/os/FileStore.cc [hunk 5]
```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
...
+      if (errno == ENXIO) {
+        break;
+      } else {
+        r = -errno;
+	return r;
+      }
     }
-    free(fiemap);
+
+    if (hole_pos > (off_t)(offset + len)) {
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: ">=" can skip extra loop

File: src/os/FileStore.cc [hunk 6]
```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
...
+  map<uint64_t, uint64_t> exomap;
+  bool fallback_to_fiemap = false;
+  FDRef fd;
+
+  int r = lfn_open(cid, oid, false, &fd);
+  if (r < 0) {
+    dout(10) << "read couldn't open " << cid << "/" << oid << ": " << cpp_strerror(r) << dendl;
+    goto done;
+  }
+
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: I think we don't want to fallback to fiemap each time if doesn't support SEEK_DATA/SEEK_HOLE. We could detect this in "init" process.

And I think we need to add config value to make sure whether use SEEK_DATA/SEEK_HOLE or not just like FIEMAP, this logic and detect logic can be decided at the beginning

**Comment by yuyuyu101** [comment]
Another concern, could we write a simple benchmark to perf "FIEMAP" and "SEEK_DATA/SEEK_HOLE" in case of this may exists orders of magnitude difference?

File: src/os/FileStore.cc (hunk 6 above)
Comment (wonzhq) [inline src/os/FileStore.cc]: > yuyuyu101 added a note 17 hours ago
I think we don't want to fallback to fiemap each time if doesn't support SEEK_DATA/SEEK_HOLE. We could detect this in "init" process.

If SEEK_DATA/SEEK_HOLE is not supported, and we don't fallback to fiemap, what's your suggestion here? Just return with a single entry in the output map, implying the file is not a sparse one?

**Comment by wonzhq** [comment]
> yuyuyu101 commented 17 hours ago
Another concern, could we write a simple benchmark to perf "FIEMAP" and "SEEK_DATA/SEEK_HOLE" in case of this may exists orders of magnitude difference?

I think the performance is not the focal point here... We want to replace the fiemap because it causes problems in some cases. And I believe the performance of SEEK_DATA/SEK_HOLE is worse than fiemap, since it needs to make multiple system calls. Maybe what we should do is to write some test cases to verify its functionality.

File: src/os/FileStore.cc (hunk 6 above)
Comment (yuyuyu101) [inline src/os/FileStore.cc]: I mean if this OS doesn't support SEEK_DATA/SEEK_HOLE, we needn't to try _do_seek_hole_data each time. We could have a flag to indicate this in init process.

**Comment by yuyuyu101** [comment]
Yeah, but for compatible reason, I think we need to make sure the performance gap between SEEK_DATA/SEEK_HOLE and FIEMAP in case of it brings too much astonished.

File: src/os/FileStore.cc (hunk 6 above)
Comment (wonzhq) [inline src/os/FileStore.cc]: The _do_seek_hole_data call is already guarded by the SEEK_DATA/SEEK_HOLE compile flag in the current code.

If you are saying the case that the code is compiled on an OS with SEEK_DATA/SEEK_HOLE support, but running on an OS without them, that's true we can detect this in the init phase.

**Comment by wonzhq** [comment]
Updated, pls review again, thx! @liewegas @yuyuyu101

**Comment by yuyuyu101** [comment]
GOOD TO ME

**Pull request merged**